skyportal/handlers/api/listing.py

```diff
@@ -0,0 +1,199 @@
+import re
+from marshmallow.exceptions import ValidationError
+from baselayer.app.access import auth_or_token
+from ..base import BaseHandler
+from ...models import (
+    DBSession,
+    User,
+    Obj,
+    Listing,
+)
+
+
+class ListingHandler(BaseHandler):
+    @auth_or_token
+    def get(self):
+        """
+        ---
+            description: get all objects corresponding to a specific user and that match a list name.
+            requestBody:
+              content:
+                application/json:
+                  schema:
+                    type: object
+                    properties:
+                      user_id:
+                        type: string
+                      list_name:
+                         type: string
+                         description: |
+                            find all objects saved to this list.
+                            If empty, will return an array of strings
+                            specifying all non-empty lists associated with this user.
+            responses:
+              200:
+                content:
+                  schema:
+                    allOf:
+                      - $ref: '#/components/schemas/Success'
+                      - type: array
+                        items:
+                          description: array of object ids or array of list names.
+        """
+        data = self.get_json()
+
+        schema = Listing.__schema__(exclude=["obj_id"])
+        try:
+            schema.load(data)
+        except ValidationError as e:
+            return self.error(f'Invalid/missing parameters: {e.normalized_messages()}')
+
+        # should we make user_id optional, defaulting to self.associated_user_object?
+        user_id = data.get("user_id")
+        if User.query.get(user_id) is None:  # verify that user exists
+            return self.error(f'User "{user_id}" does not exist!')
+
+        # verify that poster has read access to user_id's lists?
+
+        list_name = data.get("list_name")
+
+        if list_name is None:
+            listings = (
+                DBSession().query(Listing).filter(Listing.user_id == user_id)
+            ).all()
+
+            list_names = [L.list_name for L in listings]
+            list_names = list(set(list_names))  # show only unique list names
+
+            return self.success(data=list_names)
+
+        else:
+            listings = (
+                DBSession()
+                .query(Listing)
+                .filter(Listing.user_id == user_id)
+                .filter(Listing.list_name == list_name)
+            ).all()
+
+            object_ids = [L.obj_id for L in listings]
+
+            return self.success(data=object_ids)
+
+    @auth_or_token
+    def put(self):
```

Comment:
```suggestion
    def post(self):
```

Reply:
I had the impression that put is more appropriate, since you can do multiple adds to the same user+obj+list without raising an error.

Reply:
add new objects is post, merge-update objects is put